File: 5-Problem-A-Day/302. Smallest Rectangle Enclosing Black Pixels.java
class Solution {
    int top, bottom, left, right;
    int[][] dir = {{0, -1}, {0, 1}, {1, 0}, {-1, 0}};

    public int minArea(char[][] image, int x, int y) {
        top = image.length;
        left = image[0].length;
        bottom = right = -1;
        dfs(image, x, y);
        return (bottom - top + 1) * (right - left + 1);
    }

    private void dfs(char[][] image, int i, int j) {
        image[i][j] = '0';
        left = Math.min(j, left);
        right = Math.max(j, right);
        top = Math.min(i, top);
        bottom = Math.max(i, bottom);
        for (int k = 0; k < 4; k++) {
            int x = i + dir[k][0], y = j + dir[k][1];
            if (x < 0 || x >= image.length || y < 0 || y >= image[0].length || image[x][y] != '1')
                continue;
            dfs(image, x, y);
        }
    }
}

File: 2021/302. Smallest Rectangle Enclosing Black Pixels.cpp
class Solution {
public:
    int directions[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    int minArea(vector<vector<char>>& image, int x, int y) {
        int m = image.size(), n = image[0].size();
        int tX = m - 1, tY = n - 1, bX = 0, bY = 0;        
        queue<pair<int, int>> queue;
        queue.push({x, y});
        image[x][y] = 0;
        while (!queue.empty()) {
            int row = queue.front().first, col = queue.front().second;
            queue.pop();            
            tX = min(tX, row);
            tY = min(tY, col);
            bX = max(bX, row);
            bY = max(bY, col);
            for (int i = 0; i < 4; i++) {
                int newRow = row + directions[i][0], newCol = col + directions[i][1];
                if (newRow < 0 || newRow >= m || newCol < 0 || newCol >= n || image[newRow][newCol] == '0') {
                    continue;
                }
                queue.push({newRow, newCol});
                image[newRow][newCol] = '0';
            }
        }
        return (bY - tY + 1) * (bX - tX + 1);
    }
};